### PR TITLE
Block

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Add `BlockId` class to index Blocks. @gsb76 could you tell me what other operations you need for these? I expect `operator<`, etc. What else? I'd like to eventually be able to get rid of the `get_index` method.
- Add `IdPair` which is a POD that should be used when a pair between an ID and some data is necessary. The access is easier to understand because it is `id_pair.id` and `id_pair.data`.
- ~Add `BlockVector` that is a vector indexed by a `BlockId`.~

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`. For
  instructions on how to perform the CI checks locally refer to the [Dev guide
  on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run `make doc`
  to generate the documentation locally into `BUILD_DIR/docs/html`. Then open
  `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Comment by mar-celine** [comment]
In DomainHelpers I mainly use std::find, and comparison. Sometimes I use Index<Dim> and IndexIterator<Dim> to represent and iterate over block ids. I've had some time to think about this... will you consider adding more functionality to the BlockId class in the future? ElementId wraps a lot of functionality whereas BlockId currently is only standing in as a strong type.
The IdPair class seems like it will be used a lot, so I like that addition :+1:
I am less sure about having a BlockVector class - I'd want to avoid having a scenario where we have a separate iterator for every size_t we decide to make into a stronger type.
In terms of usage, do we want developers to use the stronger types in the implementation as well? Or is this only so that the types in the function signature are more meaningful? I think having clearer names are useful but I am also worried about going too far in the other direction. As a comparison, I think PairOfFaces is a good example of a type that adds clarity at a minimal cost.
@markscheel did you have any particular examples as to where wrapping these things in a meaningful name (as opposed to a raw `size_t`) would be especially helpful? Wrapping `size_t`s is something I'm starting to feel unsure about if it isn't too difficult to simply work with the `size_t`s themselves.

File: src/Domain/BlockVector.hpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
+// Distributed under the MIT License.
```
Comment (kidder) [inline src/Domain/BlockVector.hpp]: commit message an -> a

File: tests/Unit/Domain/Test_BlockVector.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+SPECTRE_TEST_CASE("Unit.Domain.BlockVector", "[Domain][Unit]") {
+  using domain::BlockId;
+  domain::BlockVector<double> block_vector{1.3, 4.8, 7.8};
+  CHECK(block_vector[BlockId{0}] == 1.3);
+  CHECK(block_vector[BlockId{1}] == 4.8);
+  CHECK(block_vector[BlockId{2}] == 7.8);
+  CHECK(block_vector == domain::BlockVector<double>{1.3, 4.8, 7.8});
+  CHECK(get_output(block_vector) == "(1.3,4.8,7.8)");
+  // Test PUP
```
Comment (kidder) [inline tests/Unit/Domain/Test_BlockVector.cpp]: test swap

Comment (kidder) [reply tests/Unit/Domain/Test_BlockVector.cpp]: test the OPs (there is a function that does this)

File: tests/Unit/Domain/Test_BlockId.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+#include "Domain/BlockId.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+SPECTRE_TEST_CASE("Unit.Domain.BlockId", "[Domain][Unit]") {
+  domain::BlockId id{10};
+  CHECK(id.get_index() == 10);
+  CHECK(id == domain::BlockId{10});
+  CHECK(id != domain::BlockId{7});
+  CHECK(get_output(id) == "[10]");
+
```
Comment (kidder) [inline tests/Unit/Domain/Test_BlockId.cpp]: test your `operators++`

File: src/DataStructures/IdPair.hpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+IdPair<std::decay_t<IdType>, std::decay_t<DataType>> make_id_pair(
+    IdType&& id, DataType&& data) {
+  return {std::forward<IdType>(id), std::forward<DataType>(data)};
+}
+
+/// \cond
+// We write the pup function as a free function to keep IdPair a POD
+// clang-tidy: no non-const references
+template <typename IdType, typename DataType>
+void pup(PUP::er& p, IdPair<IdType, DataType>& t) {  // NOLINT
```
Comment (kidder) [inline src/DataStructures/IdPair.hpp]: noexcept

File: src/DataStructures/IdPair.hpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+  p | t.id;
+  p | t.data;
+}
+
+// clang-tidy: no non-const references
+template <typename IdType, typename DataType>
+void operator|(PUP::er& p, IdPair<IdType, DataType>& t) {  // NOLINT
```
Comment (kidder) [inline src/DataStructures/IdPair.hpp]: noexcept

File: src/DataStructures/IdPair.hpp [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+ */
+template <typename IdType, typename DataType>
+struct IdPair {
+  IdType id;
+  DataType data;
+};
+
+template <typename IdType, typename DataType>
+IdPair<std::decay_t<IdType>, std::decay_t<DataType>> make_id_pair(
+    IdType&& id, DataType&& data) {
```
Comment (kidder) [inline src/DataStructures/IdPair.hpp]: noexcept

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by nilsdeppe** [comment]
@gsb76 this is designed to be a very minimal starting point. I think having the strong type in the implementation helps distinguish the various `size_t`s that represent different things, which is really what I'm going after. I think it would be good to distinguish between `BlockId`, `ElementId` and even `CornerId` at the type level to prevent accidental mixups. I know for observers we will also want to do some things at the block level, so `BlockVector` will be used there.

I think the key thing is that for *you* it might be easier to use `size_t` (at least right now), but for other people it is helpful to have different types.

Edit: @gsb76 if you can provide list of functionalities that you need I'm happy to add them to this PR :)

File: src/Domain/BlockVector.hpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  GENERATE_OP(!=)
+  GENERATE_OP(<)
+  GENERATE_OP(<=)
+  GENERATE_OP(>)
+  GENERATE_OP(>=)
+#undef GENERATE_OP
+
+  template <class S, class SAllocator = std::allocator<S>>
+  friend bool swap(const domain::BlockVector<S, SAllocator>& lhs,
+                   const domain::BlockVector<S, SAllocator>& rhs) noexcept {
```
Comment (wthrowe) [inline src/Domain/BlockVector.hpp]: I don't see how this swap function could possibly work.  It takes its arguments by const reference.  Also it returns a bool which it obtains from the result of (std::?) swap, but that returns void.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Domain/BlockVector.hpp (hunk 7 above)
Comment (nilsdeppe) [reply src/Domain/BlockVector.hpp]: It was grabbing the swap inherited from `std::vector` on line 83. I've removed this implementation but left the test.

**Comment by mar-celine** [comment]
@nilsdeppe Thanks for explaining :) I can see BlockId as something we'll add functionality to in the future, so this is a good step toward supporting that. BlockIds and ElementIds are objects that exist once the domain creation has completed - as for CornerIds, I have a few points to address:
- corner numbers are almost entirely an implementation detail except in a few cases
- they don't exist once the domain is created
- they are not used outside of DomainHelpers and the DomainCreators that use them
BlockIds persist and are being used for various things that are not related to each other, so it is good to keep track of all of those instances with a strong type.

As an aside, I have been wondering about how specifying boundaries is going to work at the DomainCreator level. Currently, the corner numbering is used to create the domain, but once the block connectivity is constructed (i.e. through BlockNeighbors and OrientationMaps), there is no way to relate this information back to the original corners. The Domain carries no information about the corners used to construct it. That means that corners can't be used after the fact to additionally specify external boundary conditions, or to identify external boundaries; it can only be done by specifying a BlockId and an Direction. With the current infrastructure the external boundary conditions would have to be set in the same algorithm that takes in the corners and returns the vector of blocks; this is the only time where both the corners and the blocks are available at the same time. Something I was considering was having the Blocks hold the CornerIds used to construct them. The corners could then be used to refer to boundaries at any later time post domain creation. Having the corners persist in the blocks might also simplify some of the algorithms in DomainHelpers. I think that the introduction of boundary conditions will bring a sufficient enough change to how DomainCreators, Blocks, and corners interact with each other that I'll want to revisit how those things are set up. I want to see how setting up boundary conditions works in a simple domain creator like Brick before I start thinking about making revisions to DomainHelpers. 

In other words, what will most convince me that we need CornerIds is if the use of corner numberings goes beyond a one-time use, like if they persist in Blocks, where they can then be used at any later time. This might find that this is the way to go once boundary conditions start being used. If that does happen to be the case, then I be happy to incorporate CornerId along with all the changes needed for boundary conditions as well.

**Comment by nilsdeppe** [comment]
Yes, the reason I think a `CornerId` would be useful is again just because it would make it easier for dumb people like me to understand all the corner magic :) 

@wthrowe Seeing this got me thinking we might wanting something like a "tagged fundamental math type". Basically a:
```
template <typename Tag, typename T>
struct StrongMathType {
  public:
    /* all the math things */
  private:
    T t_;
};
```
Then we can have anything that would normally just be a `double`, `size_t`, etc. that we want to promote to a strong type be created using:
```
struct StrongType : public StrongMathType<Tag, double> {
  using StrongMathType<Tag, double>::StrongMathType;
};
```
Then we don't have a bunch of partial implementations of wrappers, they can be added at any point. If desired we could even extend that to providing a method of disabling certain operations using type traits for specific tags.

File: tests/Unit/DataStructures/CMakeLists.txt [hunk 8]
```diff
@@ -12,6 +12,7 @@ set(LIBRARY_SOURCES
   Test_Deferred.cpp
   Test_GeneralIndexIterator.cpp
   Test_Identity.cpp
+  Test_IdPair.cpp
```
Comment (wthrowe) [inline tests/Unit/DataStructures/CMakeLists.txt]: 'P' < 'e'

File: tests/Unit/Domain/Test_BlockId.cpp [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+#include "Domain/BlockId.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+SPECTRE_TEST_CASE("Unit.Domain.BlockId", "[Domain][Unit]") {
+  domain::BlockId id{10};
+  CHECK(id.get_index() == 10);
+  CHECK(id == domain::BlockId{10});
+  CHECK(id != domain::BlockId{7});
+  CHECK(get_output(id) == "[10]");
+  id++;
```
Comment (wthrowe) [inline tests/Unit/Domain/Test_BlockId.cpp]: Check that the pre- and postfix ++ and -- operators return the correct things.

File: src/Domain/BlockVector.hpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+#include "Domain/BlockId.hpp"
+
+namespace domain {
+/*!
+ * \ingroup ComputationalDomainGroup
+ * \brief A vector that is indexed by `BlockId`.
+ */
+template <class T, class Allocator = std::allocator<T>>
+class BlockVector : private std::vector<T, Allocator> {
```
Comment (wthrowe) [inline src/Domain/BlockVector.hpp]: This class does not behave how I would expect a class with a nearly opaque index (as I believe BlockId is intended to be) to behave.  I would expect something that behaves similarly to an associative container rather than a sequence container.  `insert` seems like a strange operation for instance, since it reassociates everything later in the container with a different BlockId.

It seems that this assumes BlockId acts like a pseudo-size_t, when I would expect it to be treated as some sort of more abstract label (like ElementId or the pair used as a mortar id) (although I'm not sure how you would generate such a label for Blocks).

Since we are forced to treat BlockIds as something like size_ts anyway, it is not clear to me that this class benefits us much over just using a normal vector and indexing with `block_id.id` or some such.

File: src/Domain/BlockId.hpp [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+  // NOLINTNEXTLINE(cert-dcl21-cpp) returned object doesn't need to be const
+  BlockId operator--(int) noexcept {
+    id_--;
+    return *this;
+  }
+
+ private:
+  friend bool operator==(const BlockId& lhs, const BlockId& rhs) noexcept;
+  friend std::ostream& operator<<(std::ostream& os, const BlockId& block_id);
```
Comment (wthrowe) [inline src/Domain/BlockId.hpp]: These shouldn't be friends.  I am aware that you would like to get rid of the public interface to obtain the id, but it is easy to change them when and if that becomes possible.  If it turns out to not be possible then we don't want unnecessary friend functions.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Domain/BlockVector.hpp (hunk 10 above)
Comment (nilsdeppe) [reply src/Domain/BlockVector.hpp]: I was hoping to be able to get right of the `block_id.id` to make it extra difficult to misuse the BlockId. I certainly agree that `insert` does not make any sense in this context! Even things like `push_back` seem a bit questionable but are more reasonable than `insert`. I think what I will do is factor out BlockVector into a separate PR so we can get the rest merged and then decide whether or not we want to have every BlockId related thing go through BlockId only or whether it's possible to extract the size_t out.

Comment (kidder) [reply src/Domain/BlockVector.hpp]: yes factoring the PR into two sounds good

**Review by kidder — DISMISSED** [review]
@wthrowe please look at

File: tests/Unit/Domain/Test_BlockId.cpp [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+#include "Domain/BlockId.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+SPECTRE_TEST_CASE("Unit.Domain.BlockId", "[Domain][Unit]") {
+  domain::BlockId id{10};
+  CHECK(id.get_index() == 10);
+  CHECK(id == domain::BlockId{10});
+  CHECK(id != domain::BlockId{7});
+  CHECK(get_output(id) == "[10]");
+  CHECK(id++ == domain::BlockId{11});
```
Comment (wthrowe) [inline tests/Unit/Domain/Test_BlockId.cpp]: This is the wrong answer.  Should be 10.

File: src/Domain/BlockId.hpp [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+#include <cstddef>
+#include <ostream>
+#include <pup.h>
+
+namespace PUP {
+class er;
+}  // namespace PUP
+
+namespace domain {
```
Comment (wthrowe) [inline src/Domain/BlockId.hpp:14]: Depends on #814.

**Review by wthrowe — DISMISSED** [review]

File: src/Domain/BlockId.hpp (hunk 13 above)
Comment (nilsdeppe) [reply src/Domain/BlockId.hpp:14]: How is this possibly dependent on that? It just adds a new class...

File: tests/Unit/Domain/Test_BlockId.cpp (hunk 12 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/Test_BlockId.cpp]: Derp, done. I also added extra check statements after post increment and decrement to check that the id itself updated.

**Review by kidder — DISMISSED** [review]

**Review by mar-celine — DISMISSED** [review]

**Review by osheamonn — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
@gsb76 @osheamonn I had to rebase on master because `get_output` was moved. The change was adding `"Utilities/GetOutput.hpp"` to `Test_BlockId.cpp`

**Review by mar-celine — APPROVED** [review]

**Review by osheamonn — APPROVED** [review]

**Comment by nilsdeppe** [comment]
Travis failure is a macOS timeout.

Unblocks #755

**Comment by wthrowe** [comment]
@nilsdeppe you overrode my review without informing me, without requesting an update, and without any approval from another administrator or merger, so that you could merge your own non-critical PR.  This is a major breach of the trust you've been given as an administrator.

@teukolsky @kidder
